File: backend/api/history.py
import json
from typing import Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.database.mysql import get_db
from backend.models.record import Favorite, QARecord, RecognitionRecord
from backend.models.user import User
from backend.utils.validators import (
    sanitize_fav_type,
    sanitize_openid,
    sanitize_page_params,
    sanitize_html,
)

router = APIRouter(prefix="/api", tags=["history"])


class FavoriteRequest(BaseModel):
    openid: str
    fav_type: str
    target_id: int = 0
    title: str = ""
    content: str = ""


@router.get("/history")
def get_history(
    openid: str = Query(...),
    type: str = Query("all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    openid = sanitize_openid(openid) or openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}

    page, page_size = sanitize_page_params(page, page_size, max_page_size=100)
    limit = page_size
    offset = (page - 1) * page_size
    data = {}

    if type in ("all", "recognition"):
        query = (
            db.query(RecognitionRecord)
            .filter(RecognitionRecord.user_id == user.id)
            .order_by(RecognitionRecord.create_time.desc())
        )
        total = query.count()
        records = query.offset(offset).limit(limit).all()
        data["recognition"] = {
            "items": [
                {
                    "id": r.id,
                    "image_url": r.image_url,
                    "result": r.result,
                    "confidence": r.confidence,
                    "description": r.description,
                    "suggestion": r.suggestion,
                    "task_type": r.task_type,
                    "create_time": r.create_time.isoformat() if r.create_time else "",
                }
                for r in records
            ],
            "total": total,
            "page": page,
            "page_size": page_size,
            "has_more": (page * page_size) < total,
        }
    if type in ("all", "qa"):
        query = (
            db.query(QARecord)
            .filter(QARecord.user_id == user.id)
            .order_by(QARecord.create_time.desc())
        )
        total = query.count()
        records = query.offset(offset).limit(limit).all()
        data["qa"] = {
            "items": [
                {
                    "id": r.id,
                    "question": r.question,
                    "answer": r.answer,
                    "references": json.loads(r.references_json or "[]"),
                    "create_time": r.create_time.isoformat() if r.create_time else "",
                }
                for r in records
            ],
            "total": total,
            "page": page,
            "page_size": page_size,
            "has_more": (page * page_size) < total,
        }
    return {"code": 0, "data": data}


@router.delete("/history/recognition/{rid}")
def delete_recognition_record(
    rid: int,
    openid: str = Query(...),
    db: Session = Depends(get_db),
):
    """删除单条识别记录"""
    openid = sanitize_openid(openid) or openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}
    record = db.query(RecognitionRecord).filter(
        RecognitionRecord.id == rid, RecognitionRecord.user_id == user.id
    ).first()
    if record:
        db.delete(record)
        db.commit()
    return {"code": 0, "message": "已删除"}


@router.delete("/history/qa/{qid}")
def delete_qa_record(
    qid: int,
    openid: str = Query(...),
    db: Session = Depends(get_db),
):
    """删除单条问答记录"""
    openid = sanitize_openid(openid) or openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}
    record = db.query(QARecord).filter(
        QARecord.id == qid, QARecord.user_id == user.id
    ).first()
    if record:
        db.delete(record)
        db.commit()
    return {"code": 0, "message": "已删除"}


@router.post("/favorite")
def add_favorite(body: FavoriteRequest, db: Session = Depends(get_db)):
    openid = sanitize_openid(body.openid) or body.openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}
    fav = Favorite(
        user_id=user.id,
        fav_type=sanitize_fav_type(body.fav_type),
        target_id=body.target_id or 0,
        title=sanitize_html(body.title, max_len=256),
        content=sanitize_html(body.content, max_len=10000),
    )
    db.add(fav)
    db.commit()
    db.refresh(fav)
    return {"code": 0, "message": "收藏成功", "data": fav.id}


@router.get("/favorite")
def list_favorites(
    openid: str = Query(...),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    openid = sanitize_openid(openid) or openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}
    page, page_size = sanitize_page_params(page, page_size)
    offset = (page - 1) * page_size
    query = (
        db.query(Favorite)
        .filter(Favorite.user_id == user.id)
        .order_by(Favorite.create_time.desc())
    )
    total = query.count()
    items = query.offset(offset).limit(page_size).all()
    return {
        "code": 0,
        "data": [
            {
                "id": f.id,
                "fav_type": f.fav_type,
                "target_id": f.target_id,
                "title": f.title,
                "content": f.content,
                "create_time": f.create_time.isoformat() if f.create_time else "",
            }
            for f in items
        ],
        "total": total,
        "page": page,
        "page_size": page_size,
        "has_more": (page * page_size) < total,
    }


@router.delete("/favorite/{fid}")
def delete_favorite(
    fid: int,
    openid: str = Query(...),
    db: Session = Depends(get_db),
):
    openid = sanitize_openid(openid) or openid
    user = db.query(User).filter(User.openid == openid).first()
    if not user:
        return {"code": 401, "message": "请先登录"}
    fav = db.query(Favorite).filter(Favorite.id == fid, Favorite.user_id == user.id).first()
    if fav:
        db.delete(fav)
        db.commit()
    return {"code": 0, "message": "已取消收藏"}

File: miniapp/pages/history/history.js
/**
 * 历史记录页面 - 展示用户的识别记录和问答记录
 *
 * 功能说明：
 * 1. 标签切换：在识别记录和问答记录之间切换
 * 2. 记录列表：展示用户的历史操作记录
 * 3. 收藏问答：将问答记录添加到收藏夹
 * 4. 删除记录：删除单条识别或问答记录
 *
 * 数据流：
 * 页面显示 → 请求 API → 获取历史记录 → 渲染页面
 *
 * 依赖：
 * - request: 网络请求工具
 * - BASE_URL: 服务器基础地址（用于图片 URL 拼接）
 * - 后端 API: GET /api/history, POST /api/favorite, DELETE /api/history/*
 */

const { request, BASE_URL } = require('../../utils/request')

Page({
  /**
   * 页面数据
   *
   * tab: 当前选中的标签（'recognition' 或 'qa'）
   * recognition: 识别记录列表，每条记录包含:
   *   - id: 记录 ID
   *   - image_url: 图片地址
   *   - fullUrl: 完整图片地址（拼接 BASE_URL）
   *   - result: 识别结果
   *   - confidence: 置信度
   *   - description: 描述
   *   - suggestion: 建议
   *   - task_type: 任务类型
   *   - create_time: 创建时间
   *
   * qa: 问答记录列表，每条记录包含:
   *   - id: 记录 ID
   *   - question: 用户问题
   *   - answer: AI 回答
   *   - references: 参考资料列表
   *   - create_time: 创建时间
   *
   * loading: 是否正在加载
   */
  data: {
    tab: 'recognition',
    recognition: [],
    qa: [],
    loading: true
  },

  /**
   * 页面显示生命周期
   * 每次页面显示时加载历史记录
   */
  onShow() {
    this.loadHistory()
  },

  /**
   * 切换标签
   * @param {Object} e - 事件对象
   *   e.currentTarget.dataset.tab: 要切换的标签名
   */
  switchTab(e) {
    this.setData({ tab: e.currentTarget.dataset.tab })
  },

  /**
   * 加载历史记录
   *
   * 工作流程：
   * 1. 获取用户身份（openid）
   * 2. 验证用户已登录
   * 3. 请求历史记录 API（同时获取识别和问答记录）
   * 4. 处理图片 URL（相对路径拼接 BASE_URL）
   * 5. 更新页面数据
   */
  loadHistory() {
    const app = getApp()
    const openid = app.globalData.openid || wx.getStorageSync('openid')

    // 用户未登录，不加载
    if (!openid) {
      this.setData({ loading: false })
      return
    }

    this.setData({ loading: true })

    // 请求历史记录 API
    // type: 'all' 同时获取识别和问答记录
    request({ url: '/api/history', data: { openid, type: 'all', page: 1, page_size: 50 } })
      .then((res) => {
        if (res.code === 0) {
          // 提取识别记录和问答记录
          const recData = res.data && res.data.recognition
          const qaData = res.data && res.data.qa
          const recItems = recData && recData.items ? recData.items : []
          const qaItems = qaData && qaData.items ? qaData.items : []

          // 处理识别记录：拼接完整的图片 URL
          const recognition = recItems.map((r) => ({
            ...r,
            // 如果是相对路径（以 / 开头），拼接 BASE_URL
            fullUrl: r.image_url && r.image_url.startsWith('/') ? BASE_URL + r.image_url : r.image_url
          }))

          this.setData({ recognition, qa: qaItems, loading: false })
        } else {
          this.setData({ loading: false })
        }
      })
      .catch(() => {
        this.setData({ loading: false })
        wx.showToast({ title: '加载失败', icon: 'none' })
      })
  },

  /**
   * 收藏问答记录
   * @param {Object} e - 事件对象
   *   e.currentTarget.dataset.item: 问答记录数据
   *
   * 收藏数据：
   * - fav_type: 'qa'（问答类型）
   * - target_id: 问答记录 ID
   * - title: 用户问题（前 50 字符）
   * - content: AI 回答
   */
  favQa(e) {
    const item = e.currentTarget.dataset.item
    const app = getApp()
    const openid = app.globalData.openid || wx.getStorageSync('openid')

    // 调用收藏 API
    request({
      url: '/api/favorite',
      method: 'POST',
      data: {
        openid,
        fav_type: 'qa',
        target_id: item.id,
        title: (item.question || '').slice(0, 50),
        content: item.answer
      }
    }).then(() => {
      wx.showToast({ title: '已收藏' })
    }).catch(() => {
      wx.showToast({ title: '收藏失败', icon: 'none' })
    })
  },

  /**
   * 删除识别记录
   * @param {Object} e - 事件对象
   *   e.currentTarget.dataset.id: 记录 ID
   *
   * 工作流程：
   * 1. 弹出确认对话框
   * 2. 用户确认后调用删除 API
   * 3. 从页面数据中移除该记录
   * 4. 显示删除结果提示
   */
  deleteRecord(e) {
    const id = e.currentTarget.dataset.id

    // 弹出确认对话框
    wx.showModal({
      title: '提示',
      content: '确定删除此记录吗？',
      success: (res) => {
        if (!res.confirm) return

        const openid = getApp().globalData.openid || wx.getStorageSync('openid')

        // 调用删除 API
        request({
          url: '/api/history/recognition/' + id,
          method: 'DELETE',
          data: { openid }
        }).then(() => {
          // 从页面数据中移除该记录
          const recognition = this.data.recognition.filter(r => r.id !== id)
          this.setData({ recognition })
          wx.showToast({ title: '已删除' })
        }).catch(() => {
          wx.showToast({ title: '删除失败', icon: 'none' })
        })
      }
    })
  },

  /**
   * 删除问答记录
   * @param {Object} e - 事件对象
   *   e.currentTarget.dataset.id: 记录 ID
   *
   * 工作流程：
   * 1. 弹出确认对话框
   * 2. 用户确认后调用删除 API
   * 3. 从页面数据中移除该记录
   * 4. 显示删除结果提示
   */
  deleteQa(e) {
    const id = e.currentTarget.dataset.id

    // 弹出确认对话框
    wx.showModal({
      title: '提示',
      content: '确定删除此记录吗？',
      success: (res) => {
        if (!res.confirm) return

        const openid = getApp().globalData.openid || wx.getStorageSync('openid')

        // 调用删除 API
        request({
          url: '/api/history/qa/' + id,
          method: 'DELETE',
          data: { openid }
        }).then(() => {
          // 从页面数据中移除该记录
          const qa = this.data.qa.filter(r => r.id !== id)
          this.setData({ qa })
          wx.showToast({ title: '已删除' })
        }).catch(() => {
          wx.showToast({ title: '删除失败', icon: 'none' })
        })
      }
    })
  }
})
